### PR TITLE
Natspec: Don't copy from base function if return parameters differ

### LINKED ISSUE #10881: Solidity compiler error NatSpec multiple return value for struct in interface
## Description

Compiler throws error parsing multiple NatSpec `@return` annotations on an interface that documents the return values of what is implemented as a struct

```
Internal exception in StandardCompiler::compile: /Users/distiller/project/libsolidity/interface/Natspec.cpp(149): Throw in function static Json::Value solidity::frontend::Natspec::devDocumentation(const solidity::frontend::ContractDefinition &)
Dynamic exception type: boost::wrapexcept<solidity::langutil::InternalCompilerError>
std::exception::what:
[solidity::util::tag_comment*] =
```

## Environment

- Compiler version: 0.7.6, 0.8.0, 0.8.1
- Target EVM version (as per compiler settings): default
- Framework/IDE (e.g. Truffle or Remix): n/a
- EVM execution environment / backend / blockchain client: solidity/hardhat 2.0
- Operating system: mac os

## Steps to Reproduce
This code reproduces the issue in remix.ethereum.org

```solidity
pragma solidity =0.7.6;

interface IThing {
    /// @return x a number
    /// @return y another number
    function value() external view returns (uint128 x, uint128 y);
}

contract Thing is IThing {
    struct Value {
        uint128 x;
        uint128 y;
    }

    Value public override value;
}
```

Related, moving the annotations to the override

```
DocstringParsingError: Documentation tag "@return" is only allowed once on state-variables. /// @return x a number ^ (Relevant source part starts here and spans across multiple lines).
```

### PR DESCRIPTION
**Pull request by Marenz**
fix #10881

### CODE REVIEW

**Comment by mijovic** [comment]
Can you add a tests that has a parameters as well in a function. That can transfer to mapping in state variable. For example:
```
interface IThing {
    /// @param v value to search for
    /// @return x a number
    /// @return y another number
    function value(uint256 v) external view returns (uint128 x, uint128 y);
}

contract Thing is IThing {
    struct Value {
        uint128 x;
        uint128 y;
    }

    mapping(uint256=>Value) public override value;
}
```

What should happen with natspec comment for parameter?

**Comment by Marenz** [comment]
> What should happen with natspec comment for parameter?

In this case it wouldn't inherit at all because of a mismatch of the parameters..

**Review by mijovic — DISMISSED** [review]

**Comment by mijovic** [comment]
Only one more small question:
Should tests be in `invalid` directory, as I think they should not be there

**Review by chriseth — CHANGES_REQUESTED** [review]
Please use tests that show properly how the natspec is inherited - I actually don't know what is happening there. Also, I don't think we should have such "generic special cases". Instead, I think natspec inheritance should be handled differently for public state variables.

**Comment by Marenz** [comment]
> Also, I don't think we should have such "generic special cases". Instead, I think natspec inheritance should be handled differently for public state variables.

My code touches only the public state variable inheritance. Specifically it makes it not work when the inherited function has a different amount of return parameters... which seems a non-special generic case to me for failing inheritance (mismatch of function signature, at least from the documentation perspective, not from the ABI perspective)

> Please use tests that show properly how the natspec is inherited

As this was just a bug fix for a crash, I assumed we don't need any actual natspec test here, but I'll add one. I don't think the behavior changed really other than fixing the crash and keeping the logic of requiring matching amount of (return-)parameters..

**Comment by Marenz** [comment]
> Should tests be in invalid directory, as I think they should not be there

Well, it is invalid in terms of inheritance won't work. I guess it's not invalid in the sense that no documentation at all is generated...

**Comment by mijovic** [comment]
> > Should tests be in invalid directory, as I think they should not be there
> 
> 
> 
> Well, it is invalid in terms of inheritance won't work. I guess it's not invalid in the sense that no documentation at all is generated... 

Let's leave then where they are

File: test/libsolidity/SolidityNatspecJSON.cpp [hunk 1]
```diff
@@ -2247,6 +2247,47 @@ BOOST_AUTO_TEST_CASE(dev_return_name_no_description)
...
+			  {
+				"x": "a number",
+				"y": "another number"
+			  }
+			}
+		}
+	})ABCDEF";
+
+	char const *natspec2 = R"ABCDEF({
+		"methods": { }
```
Comment (chriseth) [inline test/libsolidity/SolidityNatspecJSON.cpp]: Can we find a solution where this is not empty? Note that if you use `FunctionType(variableDeclaration)`, you will get a function type that has the correct amount of parameters and return variables.

Comment (Marenz) [reply test/libsolidity/SolidityNatspecJSON.cpp]: Sure, the ABI function type is the same, else it couldn't override.. that's not the problem. But how would you want to map the two documented parameters to the single struct parameter for this function? How would this inherited documentation look like?

**Comment by Marenz** [comment]
@chriseth so this is updated now according to your suggestion to make it properly inherit the return parameter docu.

One thing though: All method `returns` documentation occurrences use `returns` (with an s) where as all public state variables so far used `return`. 

Example of state variable return docu:

```
		"methods" : {},
		"stateVariables" :
		{
			"state" :
			{
				"details" : "example of dev",
				"return" : "returns state"
			}
		}
	}
```
Example of a method
```
		"methods":
		{
			"g(int256)":
			{
				"returns":
				{
					"y": "value"
				}
			}
		}
```

But public variables with multiple return values now mix kind of both:

```
		"stateVariables":
		{
			"value":
			{
				"return":
				{
					"x": "a number",
					"y": "another number"
				}
			}
		}

```

Should we consistently use `returns` or should we use `return` for state variables for only one return parameter and `returns` for multiple?

**Comment by chriseth** [comment]
I would propose to use both `return` and `returns`. If there is just a single return parameter documentation, then put that inside `return` but also inside `returns`. If there are multiple, only use `returns`.

**Comment by Marenz** [comment]
But doesn't that make it more difficult for any automatic tool that wants to parse the json?

**Comment by chriseth** [comment]
If we do it as a breaking change, the tools also have to implement both, because they want to support multiple versions.

**Comment by chriseth** [comment]
We can do a breaking change with the next breaking release, though.

**Comment by Marenz** [comment]
So, just for clarification @chriseth Is your proposal only for public variables? Because functions still use `returns` even if it's multiple

**Comment by leonardoalt** [comment]
Waiting clarification from @chriseth

**Comment by chriseth** [comment]
Only backwards-compatibility for public variables.

File: Changelog.md [hunk 2]
```diff
@@ -35,6 +35,7 @@ Bugfixes:
  * SMTChecker: Fix missing type constraints on block and transaction variables in the deployment phase.
  * Type Checker: Fix internal error when override specifier is not a contract.
  * Type Checker: Make function-hash collision errors into fatal type errors.
+ * Natspec: Fix internal error when base functions has more return parameters.
```
Comment (chriseth) [inline Changelog.md]:
```suggestion
 * Natspec: Fix internal error related to the `@returns` documentation for a public state variable overriding a function.

File: test/libsolidity/SolidityNatspecJSON.cpp [hunk 3]
```diff
@@ -2309,6 +2310,63 @@ BOOST_AUTO_TEST_CASE(custom_inheritance)
...
+		}
+	})ABCDEF";
+
+	char const *natspec2 = R"ABCDEF({
+		"methods": {},
+		"stateVariables":
+		{
+			"value":
+			{
+				"return":
```
Comment (chriseth) [inline test/libsolidity/SolidityNatspecJSON.cpp]: The value for `"return"` used to be a string. I think we should not break the types.

Comment (Marenz) [reply test/libsolidity/SolidityNatspecJSON.cpp]: So in this case we would leave out `return` completely? Or how would we set the value for multiple parameters?

**Comment by leonardoalt** [comment]
Ready for review?

**Comment by Marenz** [comment]
Yes, for a while

**Review by chriseth — APPROVED** [review]

**Pull request merged**